helix-core/src/modeline.rs

```diff
@@ -0,0 +1,247 @@
+            text.lines_at(text.len_lines())
+                .reversed()
+                .take(LINES_TO_CHECK),
+        ) {
+            // can't guarantee no extra copies, since we need to regex and
+            // regexes can't operate over chunks yet, but we can at least
+            // limit how much we potentially need to copy because modelines
+            // are typically quite short.
+            if line.len_chars() > LENGTH_TO_CHECK {
+                continue;
+            }
+            let line = Cow::<str>::from(line);
+            modeline.parse_from_line(&line);
+        }
+
+        modeline
+    }
+
+    pub fn language(&self) -> Option<&str> {
+        self.language.as_deref()
+    }
+
+    pub fn indent_style(&self) -> Option<IndentStyle> {
+        self.indent_style
+    }
+
+    pub fn line_ending(&self) -> Option<LineEnding> {
+        self.line_ending
+    }
+
+    fn parse_from_line(&mut self, line: &str) {
+        let mut saw_backslash = false;
+        let split_modeline = move |c| {
+            saw_backslash = match c {
+                ':' if !saw_backslash => return true,
+                '\\' => true,
+                _ => false,
+            };
+            c == ' ' || c == '\t'
+        };
+        if let Some(pos) = MODELINE_REGEX.find(line) {
+            for option in line[pos.end()..].split(split_modeline) {
+                let parts: Vec<_> = option.split('=').collect();
+                match parts[0] {
+                    "ft" | "filetype" => {
+                        if let Some(val) = parts.get(1) {
+                            self.language = Some(vim_ft_to_helix_language(val));
+                        }
+                    }
+                    "sw" | "shiftwidth" => {
+                        if let Some(val) = parts.get(1).and_then(|val| val.parse().ok()) {
+                            if self.indent_style != Some(IndentStyle::Tabs) {
+                                self.indent_style = Some(IndentStyle::Spaces(val));
+                            }
+                        }
+                    }
+                    "ff" | "fileformat" => {
+                        if let Some(val) = parts.get(1) {
+                            self.line_ending = vim_ff_to_helix_line_ending(val);
+                        }
+                    }
+                    "noet" | "noexpandtab" => {
+                        self.indent_style = Some(IndentStyle::Tabs);
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+}
+
+fn vim_ff_to_helix_line_ending(val: &str) -> Option<LineEnding> {
+    match val {
+        "dos" => Some(LineEnding::Crlf),
+        "unix" => Some(LineEnding::LF),
+        #[cfg(feature = "unicode-lines")]
+        "mac" => Some(LineEnding::CR),
+        _ => None,
+    }
+}
+
+// if vim and helix use different names for a language, put that mapping here
+fn vim_ft_to_helix_language(val: &str) -> String {
```

Comment:
@the-mikedavis was concerned about too much vim emulation here. I tend to agree, kakoune doesn't do this either. We don't need to translate file types

Reply:
fair enough - i do kind of wonder what the right answer is for files that are intended to be shared between people using vim and helix though. maybe we do want our own format in addition?